### PR TITLE
fix: push-on-save working as expected, diagnostics cleared after successful push

### PR DESCRIPTION
**Pull request by CristiCanizales**
### What does this PR do?
Makes push on save work as expected (push every time you save) and fixes issue with push errors not being cleared from the problems panel after a successful push

### What issues does this PR fix or reference?
#4942, @ W-13749253@

### Functionality Before
- with push on save setting enabled, push on save was triggered only once, not the subsequent times
- diagnostics were kept in the problems tab even after successful push

### Functionality After
- with push on save setting enabled, push is triggered every time you save
- diagnostics are cleared from the problems tab after successful push

### CODE REVIEW

File: packages/salesforcedx-vscode-core/src/settings/pushOrDeployOnSave.ts [hunk 1]
```diff
@@ -128,6 +128,7 @@ export class DeployQueue {
           default:
             displayError(e.message);
         }
+      } finally {
```
Comment (RitamAgrawal) [inline packages/salesforcedx-vscode-core/src/settings/pushOrDeployOnSave.ts:131]: DeployQueue.locked is set to false. What does this mean?

File: packages/salesforcedx-vscode-core/test/jest/commands/deployExecutor.test.ts [hunk 2]
```diff
@@ -148,4 +161,75 @@ describe('Deploy Executor', () => {
...
+    // Assert
+    expect(getInstanceSpy).toHaveBeenCalled();
+    expect(getPackageDirectoryPathsSpy).toHaveBeenCalled();
+    expect(createOutputSpy).toHaveBeenCalled();
+    expect(appendLineSpy).toHaveBeenCalled();
+    expect(setPropertiesForFilesDeployMock).toHaveBeenCalledWith(
+      mockDeployResult
+    );
+    expect(deployRetrieveExecutorClearSpy).toHaveBeenCalled();
+    expect(sfdxCommandletExecutorClearSpy).toHaveBeenCalled();
```
Comment (RitamAgrawal) [inline packages/salesforcedx-vscode-core/test/jest/commands/deployExecutor.test.ts:215]: Is the sequence in which the error collection is cleared important?

File: packages/salesforcedx-vscode-core/test/jest/commands/deployExecutor.test.ts [hunk 3]
```diff
@@ -148,4 +161,75 @@ describe('Deploy Executor', () => {
...
+  it('should unlock queue on failure', async () => {
+    // Arrange
+    const mockUnlock = jest.fn();
+    const unlockSpy = jest
+      .spyOn(DeployQueue, 'get')
+      .mockReturnValue({ unlock: mockUnlock } as any);
+    const executor = new TestDeployExecutor(
+      'testDeploy',
+      'force_source_deploy_with_sourcepath_beta'
+    );
```
Comment (RitamAgrawal) [inline packages/salesforcedx-vscode-core/test/jest/commands/deployExecutor.test.ts:227]: Where are we specifying that we have a failure while deploying in this test?

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/baseDeployRetrieve.test.ts [hunk 4]
```diff
@@ -569,62 +569,6 @@ describe('Base Deploy Retrieve Commands', () => {
         expect(appendLineStub.calledOnce).to.equal(true);
         expect(appendLineStub.firstCall.args[0]).to.equal(expectedOutput);
       });
-
-      xit('should report any diagnostics if deploy failed', async () => {
```
Comment (RitamAgrawal) [inline packages/salesforcedx-vscode-core/test/vscode-integration/commands/baseDeployRetrieve.test.ts:573]: Are we trying to test the same behavior in 'should unlock queue on failure'?

**Review by RitamAgrawal — COMMENTED** [review]
Looks really good Cristi! I have posted few questions for better understanding.

File: packages/salesforcedx-vscode-core/src/settings/pushOrDeployOnSave.ts (hunk 1 above)
Comment (CristiCanizales) [reply packages/salesforcedx-vscode-core/src/settings/pushOrDeployOnSave.ts:131]: this unlocks the queue in order to let the subsequent pushes on save be triggered

File: packages/salesforcedx-vscode-core/test/jest/commands/deployExecutor.test.ts (hunk 3 above)
Comment (CristiCanizales) [reply packages/salesforcedx-vscode-core/test/jest/commands/deployExecutor.test.ts:227]: we are not specifying/defining the result nor the success.. thus it's a failure

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/baseDeployRetrieve.test.ts (hunk 4 above)
Comment (CristiCanizales) [reply packages/salesforcedx-vscode-core/test/vscode-integration/commands/baseDeployRetrieve.test.ts:573]: I just realized I was missing a case. Thanks!

File: packages/salesforcedx-vscode-core/test/jest/commands/deployExecutor.test.ts [hunk 5]
```diff
@@ -162,74 +163,139 @@ describe('Deploy Executor', () => {
...
+      expect(getPackageDirectoryPathsSpy).toHaveBeenCalled();
+      expect(createOutputSpy).toHaveBeenCalled();
+      expect(appendLineSpy).toHaveBeenCalled();
+      expect(setPropertiesForFilesDeployMock).toHaveBeenCalledWith(
+        mockDeployResult
+      );
+      expect(deployRetrieveExecutorClearSpy).toHaveBeenCalled();
+      expect(sfdxCommandletExecutorClearSpy).toHaveBeenCalled();
+      expect(unlockSpy).toHaveBeenCalled();
+      expect(mockUnlock).toHaveBeenCalled();
```
Comment (RitamAgrawal) [inline packages/salesforcedx-vscode-core/test/jest/commands/deployExecutor.test.ts:261]: mockUnlock is returnedValue of unlockSpy. Do we need this expect statement? Also mockUnlock should be a boolean, right?

File: packages/salesforcedx-vscode-core/test/jest/commands/deployExecutor.test.ts [hunk 6]
```diff
@@ -148,4 +162,140 @@ describe('Deploy Executor', () => {
...
+      DeployRetrieveExecutor.errorCollection
+    );
+
+    expect(handleDeployDiagnosticsSpy).toHaveBeenCalledWith(
+      mockDeployResult,
+      DeployRetrieveExecutor.errorCollection
+    );
+  });
+
+  describe('postOperation', () => {
```
Comment (RitamAgrawal) [inline packages/salesforcedx-vscode-core/test/jest/commands/deployExecutor.test.ts:194]: Separate describe block for postOperation! Good Work on keeping the separation of concerns in mind! 🎉

**Review by RitamAgrawal — APPROVED** [review]
QE Notes :
✅ With Prefer Deploy on save setting On, It runs the Deploy command after every save.
✅ Errors are disappearing after successful deploy
✅ Errors are disappearing after successful push
✅ With Prefer deploy on Save setting OFF, It runs push after every save.

**Pull request merged**